ENIGMAsystem/SHELL/Graphics_Systems/OpenGL3/GL3vertex.cpp

```diff
@@ -15,39 +17,191 @@
+    size_t size = enigma_user::vertex_get_size(buffer);
+
+    // if we haven't created a native "peer" vbo for this vertex buffer yet,
+    // then we need to do so now
+    if (it == enigma::vertexBufferPeers.end()) {
+      glGenBuffers(1, &vertexBufferPeer);
+      enigma::vertexBufferPeers[buffer] = vertexBufferPeer;
+    } else {
+      vertexBufferPeer = it->second;
+    }
+
+    bind_array_buffer(vertexBufferPeer);
+    GLint pSize;
+    glGetBufferParameteriv(GL_ARRAY_BUFFER, GL_BUFFER_SIZE, &pSize);
+
+    // if the size of the peer vbo isn't big enough to hold the new contents
+    // or vertex_freeze was called, then we need to make a call to glBufferData
+    // to allocate a bigger peer vbo or remove the GL_DYNAMIC_DRAW usage
+    if (size > pSize || vertexBuffer->frozen) {
+      GLenum usage = vertexBuffer->frozen ? GL_STATIC_DRAW : GL_DYNAMIC_DRAW;
+      glBufferData(GL_ARRAY_BUFFER, size, &vertexBuffer->vertices[0], usage);
+    } else {
+      glBufferSubData(GL_ARRAY_BUFFER, 0, size, &vertexBuffer->vertices[0]);
+    }
+
+    vertexBuffer->vertices.clear();
+    vertexBuffer->dirty = false;
+  } else {
+    bind_array_buffer(enigma::vertexBufferPeers[buffer]);
   }
 
-  bool vertex_format_exists(int id){
-    return (id >= 0 && (unsigned)id < enigma::vertexFormats.size() &&  enigma::vertexFormats[id] != nullptr);
+  //(enigma::vertexBufferPeers[buffer]);
+
+  enigma::glsl_attribute_enable_all_internal(false); //Disable all attributes
+
+  bool useTextCoords = false, useColors = false;
+  size_t offset = 0;
+  map<int,int> useCount;
+  const size_t stride = vertexFormat->stride * sizeof(float);
+  for (size_t i = 0; i < vertexFormat->flags.size(); ++i) {
+    const pair<int, int> flag = vertexFormat->flags[i];
+
+    size_t elements = 0, size = 0;
+    GLenum type = GL_FLOAT;
+    switch (flag.first) {
+      case vertex_type_float1: elements = 1; size = 1; break;
+      case vertex_type_float2: elements = 2; size = 2; break;
+      case vertex_type_float3: elements = 3; size = 3; break;
+      case vertex_type_float4: elements = 4; size = 4; break;
+      case vertex_type_color: elements = 4; size = 1; type = GL_UNSIGNED_BYTE; break;
+      case vertex_type_ubyte4: elements = 4; size = 1; type = GL_UNSIGNED_BYTE; break;
+    }
+
+    if (flag.second == vertex_usage_color) useColors = true;
+    if (flag.second == vertex_usage_textcoord) useTextCoords = true;
+
+    // NOTE: This is not what GMS1.4 does, it uses glBindAttribLocation
+    // so that in_Color0 is an alias of in_Color which glBindAttribLocation allows
+    // GMS1.4 also apparently does this binding before the shaders are linked
+    // according to a GMS tech blog:
+    // https://www.yoyogames.com/blog/16/shaders-overview-part-2
+    // https://www.yoyogames.com/blog/18/shaders-overview-part-4
+    std::string name = "";
+    int usageIndex = useCount[flag.second]++;
+    switch (flag.second) {
+      case vertex_usage_position: name = "in_Position"; break;
```

Comment:
Is there a compelling reason to use strings, here, instead of enum constants? String comparison is astoundingly slow compared to enums. If there is a reason, then put the strings in an array to be looked up by index.

Reply:
Yes, you need to read up above under the section "OpenGL3" in this pr's main comment where I explain how this works. The names are basically being treated like Direct3D "semantics" to make them compatible between the two systems. This is what makes it possible for you to use the same `vertex_format` in Direct3D and OpenGL so long as you use the name "in_Position" it will be correctly bound.